parser/src/aspects/type.rs

```diff
@@ -14,24 +15,119 @@ impl<'a> TypeAspect<'a> for Parser<'a> {
     fn parse_type(&mut self) -> ParseResult<Type<'a>> {
         self.cursor.advance(word_seps()); //consume word seps
 
-        let name_token = self.cursor.next()?;
+
+        let first_token = self.cursor.peek();
+        //if there's a parenthesis then the type is necessarily a polytype
+        let mut tpe = if first_token.token_type == RoundedLeftBracket {
+            self.parse_polytype().map(Type::Polytype)?
+        } else {
+            Type::Monotype(self.parse_monotype()?)
+        };
+
+
+        //check if there's an arrow, if some, we are maybe in a case where the type is "A => ..."
+        // (a lambda with one parameter and no parenthesis for the input, which is valid)
+        if !self.cursor.advance(word_seps().then(of_type(FatArrow))).is_some() {
+            return Ok(tpe)
+        }
+
+        if matches!(tpe, Type::Monotype(..)) {
+            tpe = Type::Polytype(Polytype {
+                inputs: vec![tpe],
+                output: Box::new(Type::Monotype(self.parse_monotype()?)),
+            })
+        }
+
+        //check if there's an arrow, if some, we are maybe in a case where the type is "A => ..."
+        // (a lambda with one parameter and no parenthesis for the input, which is valid)
+        if self.cursor.advance(word_seps().then(of_type(FatArrow))).is_some() {
+            //parse second lambda output in order to advance on the full invalid lambda expression
+            let second_rt = self.parse_monotype()?;
+            return self.expected_with("Lambda type as input of another lambda must be surrounded with parenthesis",
+                                      first_token..self.cursor.peek(),
+                                      Expected(self.unparenthesised_lambda_input_tip(tpe, second_rt)),
+            )
+        }
+
+
+        Ok(tpe)
+    }
+
+    fn parse_type_parameter_list(&mut self) -> ParseResult<Vec<Type<'a>>> {
+        self.parse_type_list(SquaredLeftBracket, SquaredRightBracket, true)
+    }
+}
+
+
+impl<'a> Parser<'a> {
+    fn unparenthesised_lambda_input_tip(&self, left_lambda: Type<'a>, lambda_out: Monotype<'a>) -> String {
+        "(".to_string() + &left_lambda.to_string() + ") => " + &lambda_out.to_string()
+    }
+
+    fn parse_polytype(&mut self) -> ParseResult<Polytype<'a>> {
+        let inputs = self.parse_type_list(RoundedLeftBracket, RoundedRightBracket, false)?;
+
+        if self.cursor.advance(word_seps().then(of_type(FatArrow))).is_none() {
+            let mut rendered_tuple = String::new();
+            rendered_tuple += "(";
+
+            if !inputs.is_empty() {
+                rendered_tuple += &format!("{}", inputs.first().unwrap());
+            }
+
+            if inputs.len() > 1 {
+                for tpe in &inputs[1..] {
+                    rendered_tuple += &format!(", {}", tpe);
+                }
+            }
```

Comment:
```suggestion
            if let Some((first, rest)) = inputs.split_first() {
                write!(rendered_tuple, "{}", first).unwrap();
                for tpe in rest {
                    write!(rendered_tuple, ", {}", tpe).unwrap();
                }
            }
```
Requires importing `use std::fmt::Write;`. It would be worthwhile to write a helper function for that.

Reply:
yeah i thought about this, why not introducing a `helper.rs` module in the parser ?

Reply:
`helper.rs` is a generic name, I was thinking of having it in the *context* crate instead, along with other functions that generate excerpts from a string or a vector.